logpyle/upgrade_db.py

```diff
@@ -0,0 +1,131 @@
+Database Upgrade Functions
+--------------------------------
+.. autofunction:: upgrade_db
+
+.. list-table:: Overview of known changes between schema versions
+   :widths: 25 25 50
+   :header-rows: 1
+
+   * - Version 1
+     - Version 2
+     - Version 3 (current)
+   * - Initial Version
+     - Added ``warnings.rank``
+     - Added ``warnings.unixtime``, ``logging``
+"""
+import logging
+import shutil
+import sqlite3
+
+logger = logging.getLogger(__name__)
+
+
+def upgrade_conn(conn: sqlite3.Connection) -> sqlite3.Connection:
+    from logpyle.runalyzer import is_gathered
+    tmp = conn.execute("select * from warnings").description
+    warning_columns = [col[0] for col in tmp]
+
+    # check if the provided connection has been gathered
+    gathered = is_gathered(conn)
+
+    # ensure that warnings table has unixtime column
+    if "unixtime" not in warning_columns:
+        print("Adding a unixtime column in the warnings table")
+        conn.execute("""
+            ALTER TABLE warnings
+                ADD unixtime integer DEFAULT NULL;
+                         """)
+
+    # ensure that warnings table has rank column
+    # nowhere to grab the rank of the process that generated
+    # the warning
+    if "rank" not in warning_columns:
+        print("Adding a rank column in the warnings table")
+        conn.execute("""
+            ALTER TABLE warnings
+                ADD rank integer DEFAULT NULL;
+                         """)
+
+    tables = [col[0] for col in conn.execute("""
+                    SELECT name
+                    FROM sqlite_master
+                    WHERE type='table'
+                       """)]
+
+    print("Ensuring a logging table exists")
+    if "logging" not in tables:
+        conn.execute("""
+            CREATE TABLE logging (
+                rank integer,
+                step integer,
+                unixtime integer,
+                level text,
+                message text,
+                filename text,
+                lineno integer
+                )""")
+        if gathered:
+            conn.execute("""
+                ALTER TABLE logging
+                ADD run_id integer;
+                             """)
+
+    return conn
+
+
+def upgrade_db(
+        dbfile: str, suffix: str, overwrite: bool
+        ) -> None:
+    """
+    Upgrade a database file to the most recent format. If the
+    `overwrite` parameter is True, it simply modifies the existing
+    database and uses the same file name for the upgraded database.
+    Otherwise, a new database is created with a separate filename
+    by appending the given suffix to the original file's base name
+    using `filename + suffix + "." + file_ext`.
+
+    Parameters
+    ----------
+    dbfile
+      A database file path
+
+    suffix
+      a suffix to be appended to the filename for the
+      upgraded database
+
+    overwrite
+      a boolean value indicating
+      whether to overwrite the original database or not
+    """
```

Comment:
Add a note that it (currently) updates all versions to v3, with v1 untested.

Reply:
Added a Note in the auto docs at the top of the file